registry/handlers/catalog.go

```diff
@@ -0,0 +1,56 @@
+package handlers
+
+import (
+	"encoding/json"
+	"net/http"
+	"strconv"
+
+	"github.com/docker/distribution/registry/api/errcode"
+	"github.com/docker/distribution/registry/storage"
+	"github.com/gorilla/handlers"
+)
+
+const maximumReturnedEntries = 100
+
+func catalogDispatcher(ctx *Context, r *http.Request) http.Handler {
+	catalogHandler := &catalogHandler{
+		Context: ctx,
+	}
+
+	return handlers.MethodHandler{
+		"GET": http.HandlerFunc(catalogHandler.GetCatalog),
+	}
+}
+
+type catalogHandler struct {
+	*Context
+}
+
+type catalogAPIResponse struct {
+	Repositories []string `json:"repositories"`
+}
+
+func (ch *catalogHandler) GetCatalog(w http.ResponseWriter, r *http.Request) {
+	q := r.URL.Query()
+	lastEntry := q.Get("last")
+	maxEntries, err := strconv.Atoi(q.Get("n"))
+	if err != nil || maxEntries < 0 {
+		maxEntries = maximumReturnedEntries
+	}
+
+	repos, err := storage.GetRepositories(ch, ch.App.driver, maxEntries, lastEntry)
```

Comment:
I don't have enough time today to give you full details, but the handlers should not be directly making calls to storage. We'll need a new API method on `distribution.Namespace` to provide a catalog object. Something like the following:

```
Catalog() CatalogService

type CatalogService interface {
    // Repositories fills p with up to len(p) entries, beginning after the provided q parameter.
    Repositories(p []string, q string)  (n int, err error)
}
```

May want to generalize the interpretation of `q` to allow more interesting queries against the catalog in the future. Note that this is reentrant, so if you have a max value of `n` entries, `p` is filled with those entries, similar to the way readers work.